### PR TITLE
Made Phase methods functional

### LINKED ISSUE #254: Remove in/out ContactGoals @cp_goals
WalkCycle currently modifies its input argument cp_goals. It is super-confusing: re-factor by updating in caller.

### PR DESCRIPTION
**Pull request by dellaert**
Added `Phase::ContactPointGoals` type and made `Phase::contactPointObjectives` functional, moving updating of goal points to `Phase::updateContactPointGoals`.
Closes #254

### CODE REVIEW

File: gtdynamics/utils/Phase.cpp [hunk 1]
```diff
@@ -39,17 +39,17 @@ void Phase::print(const string &s) const {
...
   for (auto &&kv : all_contact_points) {
     const string &name = kv.first;
-    Point3 &cp_goal = cp_goals->at(name);
+    const Point3 &cp_goal = cp_goals.at(name);
     const bool stance = hasContact(name);
     auto goal_trajectory =
         stance ? StanceTrajectory(cp_goal, num_time_steps_)
                : SimpleSwingTrajectory(cp_goal, step, num_time_steps_);
-    if (!stance) cp_goal += step;  // Update the goal if swing
+    // if (!stance) cp_goal += step;  // Update the goal if swing
```
Comment (dellaert) [inline gtdynamics/utils/Phase.cpp:52]: remove

File: gtdynamics/utils/Phase.cpp [hunk 2]
```diff
@@ -58,6 +58,20 @@ NonlinearFactorGraph Phase::contactPointObjectives(
   return factors;
 }
 
+Phase::ContactPointGoals Phase::updateContactPointGoals(
+    const ContactPoints &all_contact_points, const Point3 &step,
+    const ContactPointGoals &cp_goals) const {
+  ContactPointGoals new_goals;
+
+  for (auto &&kv : all_contact_points) {
```
Comment (dellaert) [inline gtdynamics/utils/Phase.cpp:66]: add doc

**Review by danbarla — APPROVED** [review]
I feel honored to review :)
just a thought - maybe we want to test for cp_goals update also in testWalkCycle.cpp?
since we update cp_goals in the loop (in WalkCycle.cpp) after the factors of the phase are added.
at the end cp_goals should be the ContactPointsGoals of the last phase?
or maybe we don't care?

**Comment by dellaert** [comment]
> I feel honored to review :)
> just a thought - maybe we want to test for cp_goals update also in testWalkCycle.cpp?
> since we update cp_goals in the loop (in WalkCycle.cpp) after the factors of the phase are added.
> at the end cp_goals should be the ContactPointsGoals of the last phase?
> or maybe we don't care?

Good suggestion!

**Pull request merged**